viewer-prototype/src/browser/trace-viewer/components/timegraph-output-component.tsx

```diff
@@ -105,19 +109,36 @@ export class TimegraphOutputComponent extends AbstractTreeOutputComponent<Timegr
                 timegraphTree: treeResponse.model.entries,
                 styleModel: styleResponse.model
             });
+            this.chartLayer.updateChart();
         }
+
+        if (this.state.collapsedNodes !== _prevState.collapsedNodes && this.chartLayer) {
+            this.chartLayer.updateChart();
+        }
+    }
+
+    private onCollapse(id: number) {
+        let newList = [...this.state.collapsedNodes];
+
+        const exist = this.state.collapsedNodes.find(expandId => expandId === id);
+
+        if (exist !== undefined) {
+            newList = newList.filter(collapsed => id !== collapsed);
+        } else {
+            newList = newList.concat(id);
+        }
+        this.setState({ collapsedNodes: newList });
     }
 
     renderTree(): React.ReactNode {
-        return <React.Fragment>
-            {this.state.timegraphTree.map((entry, i) => {
-                if (entry.parentId !== -1) {
-                    return <p style={{ height: this.props.style.rowHeight, margin: 0 }} key={i}>
-                        {entry.labels[0] + '\n'}
-                    </p>;
-                }
-            })}
-        </React.Fragment>;
+        this.onCollapse = this.onCollapse.bind(this);
```

Comment:
This line can go in the constructor, so it is done only once instead of for every render